features/305-microservice/microservice.md

```diff
@@ -83,3 +119,11 @@ In the future, it is possible to have configurable replicas or even auto horizon
 ## Routing to the microservice
 
 The existing routing mechanism for cloud code is enhanced to support routing to microservices.
+
+## Preconfigured environment
+
+To avoid maintain too many minor versions, only major versions will be
+provided. For example, if user set `running_environment` to `nodejs` and
+`environment_version` to `10`, docker base image `node:10` will be used. So the
```

Comment:
So for Python, we only offer two variants for 2.x and 3.x? I think the offered environment version should depends on the runtime/tooling, since not all projects follow semver.

Reply:
Yes, you are right. For the python case, we may want to offer 2.7, 3.5, 3.6, 3.7 and 3.8? And for node maybe 8, 10 and 12. 

In the current design, every `running_environment` and `environment_version` will have 1 dockerfile. So we can decide what versions we want to support for each language.

Reply:
> For the python case, we may want to offer 2.7, 3.5, 3.6, 3.7 and 3.8? And for node maybe 8, 10 and 12.

That sounds reasonable to me. If developers want to pin a specific version (e.g. node 12.9.1), they can write their own Dockerfile.

> In the current design, every running_environment and environment_version will have 1 dockerfile. So we can decide what versions we want to support for each language.

Would we support a docker-like image spec, instead of two separate fields? e.g. `runtime: python:2.7`, `runtime: node:10`, etc.

Reply:
> That sounds reasonable to me. If developers want to pin a specific version (e.g. node 12.9.1), they can write their own Dockerfile.

🙌🏻

> Would we support a docker-like image spec, instead of two separate fields? e.g. runtime: python:2.7, runtime: node:10, etc.

If we use docker-like image spec, will user think they can use any docker images in docker hub? Actually, I think the word `running_environment` is too long, I used it because we used it in cloud function. Given that we may want to change cloud function, maybe we don't have to use the same word. I love the word `runtime`, how about using `runtime` and `runtime_version`?

Reply:
For compiled language like Golang, the word runtime may not be appropriate. Travis uses the word `language`.

Reply:
After discussion, we will use the term `template`. e.g. `template: nodejs:10`